tests/staggered_gsmear_test_utils.h

```diff
@@ -0,0 +1,405 @@
+    initExtendedField(qdp_inlink_ex, qdp_inlink);
+    // Prepare two link field:
+    if(verify_results or (gtest_type == gsmear_test_type::GaussianSmear and  compute_two_link == false)) computeTwoLinkCPU(qdp_twolnk, qdp_inlink_ex, &gauge_param);
+
+    // Reorder gauge fields to MILC order
+    if(compute_two_link or gtest_type == gsmear_test_type::TwoLink)
+      reorderQDPtoMILC(milc_inlink, qdp_inlink, V, gauge_site_size, gauge_param.cpu_prec, gauge_param.cpu_prec);
+    else
+      reorderQDPtoMILC(milc_inlink, qdp_twolnk, V, gauge_site_size, gauge_param.cpu_prec, gauge_param.cpu_prec);
+
+    if ( gtest_type == gsmear_test_type::GaussianSmear ) {
+      // Specific gauge parameters for MILC
+      int link_pad = 3 * gauge_param.ga_pad;
+
+      gauge_param.reconstruct                   = QUDA_RECONSTRUCT_NO;
+      gauge_param.reconstruct_sloppy            = QUDA_RECONSTRUCT_NO;
+      gauge_param.reconstruct_refinement_sloppy = QUDA_RECONSTRUCT_NO;
+
+      gauge_param.type                 = QUDA_ASQTAD_LONG_LINKS;
+      gauge_param.ga_pad               = link_pad;
+      gauge_param.staggered_phase_type = QUDA_STAGGERED_PHASE_NO;
+    }
+    //
+    loadGaugeQuda(milc_inlink, &gauge_param);
+    //  
+    if(gtest_type == gsmear_test_type::GaussianSmear) {  
+      // Create ghost gauge fields in case of multi GPU builds.
+      reorderQDPtoMILC(milc_twolnk, qdp_twolnk, V, gauge_site_size, gauge_param.cpu_prec, gauge_param.cpu_prec);
+      //
+      gauge_param.type     = QUDA_ASQTAD_LONG_LINKS;
+      gauge_param.location = QUDA_CPU_FIELD_LOCATION;
+
+      GaugeFieldParam cpuTwoLinkParam(gauge_param, milc_twolnk);
+      cpuTwoLinkParam.ghostExchange = QUDA_GHOST_EXCHANGE_PAD;
+      cpuTwoLink                    = GaugeField::Create(cpuTwoLinkParam);    
+    
+      ColorSpinorParam cs_param;
+
+      constructStaggeredTestSpinorParam(&cs_param, &inv_param, &gauge_param);
+
+      spinor    = std::make_unique<ColorSpinorField>(cs_param);
+      spinorRef = std::make_unique<ColorSpinorField>(cs_param);
+      tmp       = std::make_unique<ColorSpinorField>(cs_param);
+      tmp2      = std::make_unique<ColorSpinorField>(cs_param);    
+
+      spinor->Source(QUDA_RANDOM_SOURCE);
+    
+      *tmp = *spinor;
+    }
+  }
+
+  void end()
+  {
+    // Clean up gauge fields
+    for (int dir = 0; dir < 4; dir++) {
+      host_free(qdp_inlink[dir]);
+      host_free(qdp_inlink_ex[dir]);
+      host_free(qdp_twolnk[dir]);
+      host_free(qdp_ref_twolnk[dir]);
+    }
+  
+    host_free(milc_inlink);
+    host_free(milc_twolnk);
+
+    if (cpuTwoLink != nullptr) { delete cpuTwoLink; cpuTwoLink = nullptr; }
+
+    if(gtest_type == gsmear_test_type::GaussianSmear) {
+      tmp2.reset();
+      tmp.reset();
+      spinor.reset();
+      spinorRef.reset();
+    }
+
+    freeGaugeQuda();
+
+    commDimPartitionedReset();
+  }
+
+  GSmearTime gsmearCUDA(int niter) //niter
```

Comment:
Probably this function should not be called `gsmearCUDA` since QUDA runs on more than CUDA nowadays 😉

Reply:
sure, done.